[XRay] Only avoid outlining pseudo-instructions, not whole blocks

dhoekwater opened the pull request:
While some pseudo-instructions, such as fentry and PATCHABLE
instructions, must not be outlined, safe instruction sequences
within the same basic block can be outlined. Move the logic that
prevents pseudo-instruction outlining from the basic block level
to the instruction level. Doing so allows more instructions to be
safely outlined.

dhoekwater commented:
Previously approved in [D156354](https://reviews.llvm.org/D156354).

MaskRay commented on llvm/test/CodeGen/RISCV/machine-outliner-patchable.ll:
```diff
@@ -75,3 +75,12 @@ if.end:
+; CHECK-NEXT:    li      a1, 2
+; CHECK-NEXT:    li      a2, 3
+; CHECK-NEXT:    li      a3, 4
+; CHECK-NEXT:    jr      t0
```
ensure '\n' at the end of the file

dhoekwater committed:
[XRay] Only avoid outlining pseudo-instructions, not whole blocks

While some pseudo-instructions, such as fentry and PATCHABLE
instructions, must not be outlined, safe instruction sequences
within the same basic block can be outlined. Move the logic that
prevents pseudo-instruction outlining from the basic block level
to the instruction level. Doing so allows more instructions to be
safely outlined.

Pull request: https://github.com/llvm/llvm-project/pull/76520
